tests/utils/streams/test_kafka.py

```diff
@@ -17,268 +18,147 @@
     KafkaProducer,
     as_kafka_configuration_bool,
 )
-from snuba.utils.streams.types import (
-    Message,
-    Partition,
-    Topic,
-)
+from snuba.utils.streams.types import Message, Partition, Topic
+from tests.utils.streams.mixins import StreamsTestMixin
 from tests.backends.confluent_kafka import FakeConfluentKafkaProducer
 
 
-configuration = {"bootstrap.servers": "127.0.0.1"}
+class TestCodec(Codec[KafkaPayload, int]):
+    def encode(self, value: int) -> KafkaPayload:
+        return KafkaPayload(None, f"{value}".encode("utf-8"))
 
-
-@pytest.yield_fixture
-def topic() -> Iterator[Topic]:
-    name = f"test-{uuid.uuid1().hex}"
-    client = AdminClient(configuration)
-    [[key, future]] = client.create_topics(
-        [NewTopic(name, num_partitions=1, replication_factor=1)]
-    ).items()
-    assert key == name
-    assert future.result() is None
-    try:
-        yield Topic(name)
-    finally:
-        [[key, future]] = client.delete_topics([name]).items()
-        assert key == name
-        assert future.result() is None
+    def decode(self, value: KafkaPayload) -> int:
+        return int(value.value.decode("utf-8"))
 
 
-def build_producer() -> KafkaProducer[KafkaPayload]:
-    codec: PassthroughCodec[KafkaPayload] = PassthroughCodec()
-    return KafkaProducer(configuration, codec=codec)
+class KafkaStreamsTestCase(StreamsTestMixin, TestCase):
 
+    configuration = {"bootstrap.servers": "127.0.0.1"}
+    codec = TestCodec()
 
-def test_consumer_backend(topic: Topic) -> None:
-    def build_consumer() -> Consumer[KafkaPayload]:
-        codec: PassthroughCodec[KafkaPayload] = PassthroughCodec()
+    @contextlib.contextmanager
+    def get_topic(self) -> Iterator[Topic]:
+        name = f"test-{uuid.uuid1().hex}"
+        client = AdminClient(self.configuration)
+        [[key, future]] = client.create_topics(
+            [NewTopic(name, num_partitions=1, replication_factor=1)]
+        ).items()
+        assert key == name
+        assert future.result() is None
+        try:
+            yield Topic(name)
+        finally:
+            [[key, future]] = client.delete_topics([name]).items()
+            assert key == name
+            assert future.result() is None
+
+    def get_consumer(
+        self, group: str, auto_offset_reset: str = "earliest"
+    ) -> KafkaConsumer[int]:
```

Comment:
mypy surprisingly doesn't mind this, I'm not sure if the multiple inheritance is tripping it up or if it actually considers this a valid signature for a subclass. I'm not sure what approach makes sense to making `auto_offset_reset` a generic parameter so I'm continuing to kick that can down the road.

Reply:
Maybe it is happy because of the default value ? that makes it factually substitutable to the parent class method ?

Reply:
I feel like I've seen it cause type errors if the subclass doesn't exactly match the superclass, but that must have been a different context (or I am misremembering… I swore I even saw it cause errors if the return type was a subtype of the return type of the superclass (e.g. `-> Consumer[int]` on the supertype here, `-> KafkaConsumer[int]` on the subtype.)

I think you're right though — if I remove the default, it *does* error, so it's obviously checking to make sure that the signature is at least compatible in both cases, even if it's not exactly the same. Pretty cool.